extensions/amp-viewer-integration/0.1/findtext.js

```diff
@@ -0,0 +1,451 @@
+          r.start.offset -= end.offset;
+        } else {
+          break;
+        }
+        if (end.node == r.end.node) {
+          r.end.node = newText;
+          r.end.offset -= end.offset;
+        } else {
+          break;
+        }
+      }
+      return;
+    }
+    const next = nextTextNode(start.node);
+    markSingleTextNode(
+        start.node, start.offset, start.node.wholeText.length, marked);
+    if (!next) {
+      break;
+    }
+    start = new TextPos(next, 0);
+  }
+};
+
+/**
+ * Wraps a text range [start, end) in a single text node.
+ * Returns a text node for the suffix text if it exists.
+ * @param {!Text} node
+ * @param {number} start
+ * @param {number} end
+ * @return {?Text}
+ * @param {!Array<!Element>} marked
+ */
+const markSingleTextNode = function(node, start, end, marked) {
+  if (start >= end) {
+    // Do nothing
+    return null;
+  }
+  const parent = node.parentNode;
+  const text = node.wholeText;
+  if (start > 0) {
+    parent.insertBefore(document.createTextNode(
+        text.substring(0, start)), node);
+  }
+  const span = document.createElement('span');
+  span.appendChild(document.createTextNode(text.substring(start, end)));
+  parent.insertBefore(span, node);
+  marked.push(span);
+
+  let endText = null;
+  if (end < text.length) {
+    endText = document.createTextNode(text.substring(end));
+    parent.insertBefore(endText, node);
+  }
+  parent.removeChild(node);
+  return endText;
+};
+
+
+/**
+ * nextTextNode finds the next sibling text node or
+ *   the next text node in the siblings of the parent.
+ * @param {!Text} textNode The node to start to find the next text node.
+ * @return {?Text}
+ */
+const nextTextNode = function(textNode) {
+  // If leaving is true, find the next node from siblings or the parent.
+  // If leaving is false, find the next node from childrens.
+  let leaving = true;
+  /** @type {Node} */
+  let node = textNode;
+  while (true) {
+    if (node == null) {
+      // No more parent.
+      return null;
+    }
+    if (leaving) {
+      const next = node.nextSibling;
+      if (next) {
+        node = next;
+        // visits childrens of node.
+        leaving = false;
+      } else {
+        // back to the parent.
+        node = node.parentNode;
+      }
+      continue;
+    }
+    if (node instanceof Text) {
+      return node;
+    }
+    if (!node.firstChild) {
+      leaving = true;
+    } else {
+      node = node.firstChild;
+    }
+  }
+};
+
+
+/**
+ * A special TextPos object to represent a whitespace injected
+ *   between two block nodes.
+ * @type {!TextPos}
+ */
+const posDomDelimiter = new TextPos(document.createTextNode(' '), 0);
+
+/**
+ * TextScanner visits text nodes under a root node and
+ *   returns charcter positions respectively.
+ */
+export class TextScanner {
+  /**
+   * @param {!Node} root The root node to visit.
+   */
+  constructor(root) {
+    this.internal_ = new TextScannerInternal(root, false);
```

Comment:
Is this composition necessary? Seems equivalent to having a flat object with private methods?